cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -0,0 +1,381 @@
+      }
+      return streamRDD;
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @Override
+  public <T> T getOriginalSparkContext() {
+    return getApacheSparkContext().getContext();
+  }
+
+  @Override
+  public synchronized <T extends Dataset> T getDataset(String name, Map<String, String> arguments) {
+    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, name, getRuntimeArguments());
+    datasetArgs.putAll(arguments);
+
+    String key = name + ImmutableSortedMap.copyOf(datasetArgs).toString();
+
+    @SuppressWarnings("unchecked")
+    T dataset = (T) datasets.get(key);
+    if (dataset == null) {
+      dataset = createDataset(name, datasetArgs);
+      datasets.put(key, dataset);
+    }
+    return dataset;
+  }
+
+  @Override
+  public synchronized void close() {
+    if (stopped) {
+      return;
+    }
+    stopped = true;
+    if (apacheSparkContext != null) {
+      apacheSparkContext.stop();
+    }
+    // No need to flush datasets, just close them
+    for (Dataset dataset : datasets.values()) {
+      Closeables.closeQuietly(dataset);
+    }
+  }
+
+  /**
+   * For all datasets created through {@link #getDataset(String, Map)} so far that implements
+   * {@link TransactionAware}, call {@link TransactionAware#commitTx()}.
+   */
+  public void flushDatasets() throws Exception {
+    for (Dataset dataset : datasets.values()) {
+      if (dataset instanceof TransactionAware) {
+        ((TransactionAware) dataset).commitTx();
+      }
+    }
+  }
+
+  /**
+   * Returns true if this context is closed already.
+   */
+  public synchronized boolean isStopped() {
+    return stopped;
+  }
+
+  /**
+   * Returns a {@link BatchReadable} that is implemented by the given dataset. This method is expected to be
+   * called by {@link SparkDatasetInputFormat}.
+   *
+   * @param datasetName name of the dataset
+   * @param arguments arguments for the dataset
+   * @param <K> key type for the BatchReadable
+   * @param <V> value type for the BatchReadable
+   * @return the BatchReadable
+   * @throws DatasetInstantiationException if cannot find the dataset
+   * @throws IllegalArgumentException if the dataset does not implement BatchReadable
+   */
+  public <K, V> BatchReadable<K, V> getBatchReadable(String datasetName, Map<String, String> arguments) {
+    final Dataset dataset = createDataset(datasetName, arguments);
+    Preconditions.checkArgument(dataset instanceof BatchReadable, "Dataset %s of type %s does not implements %s",
+                                datasetName, dataset.getClass().getName(), BatchReadable.class.getName());
+
+    @SuppressWarnings("unchecked")
+    final BatchReadable<K, V> delegate = (BatchReadable<K, V>) dataset;
+    return new BatchReadable<K, V>() {
+      @Override
+      public List<Split> getSplits() {
+        try {
+          return delegate.getSplits();
+        } finally {
+          commitAndClose(dataset);
```

Comment:
In general, anybody using a BatchReadable can call getSplits at any time. But if this particular BatchReadable is only used for passing it to DatasetInputFormat, with in turn is only called by Spark in the way you describe, then it is ok.